src/armature/Generator.ts

```diff
@@ -1,38 +1,189 @@
 import { RandomGenerator } from '../utils/random';
+import { Model } from './Model';
 import { Node, Point } from './Node';
 
-import { range } from 'lodash';
+import { minBy, range } from 'lodash';
+
+/**
+ * A function to generate a component from a spawn point. These are the body of rule definitions.
+ */
+type GeneratorFn = (root: Point, instance: GeneratorInstance) => void;
 
 type Definition = {
     weight: number;
-    generator(root: Point);
+    generator: GeneratorFn;
 };
 
 type SpawnPoint = {
     component: string;
     at: Point;
 };
 
+/**
+ * A cost function returns the cost of an entire model. The new nodes that were added
+ * since the last iteration of generation are passed to the cost function so that, if you can,
+ * you can just compute the incremental cost and add it to the previous cost, `instance.getCost()`.
+ */
+export type CostFn = (instance: GeneratorInstance, nodes: Node[]) => number;
```

Comment:
Unfortunately because `CostFn` has a `GeneratorInstance` parameter, if I defined this in `CostFunctions.ts`, it would need to first import `Generator.ts` to know what that is, but `Generator.ts` needs to know what `CostFn` is, so it would either have to redefine it or import from `CostFunctions.ts`, which is a circular dependency. So basically it kind of needs to be in the same file as `GeneratorInstance` is defined because import rules are weird

Reply:
Ugh javascript